### PR TITLE
[DRAFT] New Paywall Contract (DAP-4287)

### PR DESCRIPTION
**Pull request by alsakhaev**

### CODE REVIEW

File: contract/src/lib.rs [hunk 1]
```diff
@@ -1,19 +1,70 @@
...
+    {
+        let mut s = serializer.serialize_struct("PaidContent", 5)?;
+        s.serialize_field("id", &self.id)?;
+        s.serialize_field("link", &self.link)?;
+        s.serialize_field("author", &self.author)?;
+        s.serialize_field("cost", &self.cost)?;
+        s.serialize_field("context_id", &self.context_id)?;
+        s.end()
+    }
+}
```
Comment (alsakhaev) [inline contract/src/lib.rs]: it is not necessary. 
`#[near_bindgen]` should generate it

File: contract/src/lib.rs [hunk 2]
```diff
@@ -1,19 +1,70 @@
...
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        let mut s = serializer.serialize_struct("ReturningContent", 2)?;
+        s.serialize_field("content", &self.content)?;
+        s.serialize_field("is_purchased", &self.is_purchased)?;
+        s.end()
+    }
+}
```
Comment (alsakhaev) [inline contract/src/lib.rs]: it is not necessary. 
`#[near_bindgen]` should generate it

File: contract/src/lib.rs [hunk 3]
```diff
@@ -22,60 +73,115 @@ impl Default for Contract {
...
-            self.purchases_per_account
-                .get(&account_id)
-                .unwrap_or_else(|| {
-                    UnorderedSet::new(StorageKeys::PurchasesPerAccountInner {
-                        account_hash: env::sha256_array(account_id.as_bytes()),
-                    })
-                });
+    pub fn add_paid_content(&mut self, link: String, cost: String, context_id: String) -> String {
+        let author = env::predecessor_account_id();
+        let id = calculate_hash(&(context_id.clone() + &link)).to_string();
```
Comment (alsakhaev) [inline contract/src/lib.rs]: Values `{ context_id: '123', link: '456' }` and `{ context_id: '1', link: '23456' }` will produce the same hash.

File: contract/src/lib.rs [hunk 4]
```diff
@@ -85,119 +191,193 @@ impl Contract {
                     })
                 });
 
+        // Check if the content has already been purchased
+        assert!(
+            !content_purchases.contains(&id),
+            "Content already purchased"
+        );
+
+        let x = self.paid_contents.get(&id).unwrap();
```
Comment (alsakhaev) [inline contract/src/lib.rs]: Rename `x` to `content` please

File: contract/src/lib.rs [hunk 5]
```diff
@@ -85,119 +191,193 @@ impl Contract {
...
+        );
+
         // Add the content to the purchased list for the account
-        content_purchases.clear();
+        content_purchases.insert(&id);
         self.purchases_per_account
             .insert(&account_id, &content_purchases);
+
+        // Transfer attached deposit to the author
+        Promise::new(x.author).transfer(env::attached_deposit())
```
Comment (alsakhaev) [inline contract/src/lib.rs]: The transfer will be performed in the next block. Can it fail? The content is purchased, but no tokens are transferred, for example.

File: contract/src/lib.rs (hunk 1 above)
Comment (Ni-2) [reply contract/src/lib.rs]: fixed

File: contract/src/lib.rs (hunk 2 above)
Comment (Ni-2) [reply contract/src/lib.rs]: fixed

File: contract/src/lib.rs (hunk 4 above)
Comment (Ni-2) [reply contract/src/lib.rs]: fixed

File: contract/src/lib.rs (hunk 5 above)
Comment (Ni-2) [reply contract/src/lib.rs]: In the `test_buy_content` test I check if the number of buyer tokens decreases by the amount of the cost. I didn't find the ability to check if the author get these tokens 🤷

File: contract/src/lib.rs (hunk 3 above)
Comment (Ni-2) [reply contract/src/lib.rs]: fixed

**Pull request merged**